tools/autograd/gen_variable_type.py

```diff
@@ -600,6 +666,28 @@ def wrap_view_single(output_var, base_var):
         else:
             return 'as_variable({})'.format(call), []
 
+    def enforce_same_tensorimpl_and_storage(env, call):
+        pre_call_block = ''
+        post_call_block = ''
+        if declaration['name'] not in DONT_ENFORCE_SAME_TENSOR_IMPL_OR_STORAGE:
+            for arg in env.get('unpacked_args', []):
+                dynamic_type = env['unpacked_args_dynamic_type'][arg]
+                if dynamic_type == 'TensorList':
+                    pre_call_block += SAVE_TENSORLIST_STORAGE.substitute(tensorlist_name=arg)
+                    pre_call_block += SAVE_TENSORLIST_IMPL.substitute(tensorlist_name=arg)
+                    post_call_block += ENFORCE_SAME_TENSORLIST_STORAGE.substitute(tensorlist_name=arg)
+                    post_call_block += ENFORCE_SAME_TENSORLIST_IMPL.substitute(tensorlist_name=arg)
+                elif dynamic_type in ['IntegerTensor', 'IndexTensor', 'BoolTensor', 'Tensor']:
+                    pre_call_block += SAVE_TENSOR_STORAGE.substitute(tensor_name=arg)
+                    pre_call_block += SAVE_TENSOR_IMPL.substitute(tensor_name=arg)
+                    post_call_block += ENFORCE_SAME_TENSOR_STORAGE.substitute(tensor_name=arg)
+                    post_call_block += ENFORCE_SAME_TENSOR_IMPL.substitute(tensor_name=arg)
+        if pre_call_block:
```

Comment:
this looks a little strange, because if there's a pre_call_block there has to be a post_call_block right?

I'd expect something like an assert that they are either both set or neither set and a single if check.